### ISSUE TITLE
Fall Through filter subscription

### ISSUE THREAD

**Issue by mikewo** [Azure/azure-service-bus#277]
# Description

## Actual Behavior
1. By default if you deliver a message to a topic that doesn't have a subscription, or there are filtered subscriptions that don't end up getting the message because it doesn't match the filters, the message essentially just goes away. No one cared about it, which makes sense for the message to then just disappear.

## Expected Behavior
1.  It would be nice to be able to actually create a type of subscription that was to catch these types of messages in order to determine if there is mis-configuration in the system. The goal would be to catch if someone starts sending messages to a topic that no one is paying attention to that type of message, which could be a flag of mis-configuration somewhere. I know there exists the EnableFilteringMessagesBeforePublishing which may be useful in catching these scenarios in development, but I'm also interested in catching them in production as well.
2. I image this is a special type of Filter you would set when you create the subscription (a peer to Correlation, SQLFilter, etc.) that the server would see and then if a message wasn't delivered to any other subscription a copy of it would be placed in this last subscription. 
3. The system would only allow one of these fall through filters on a topic at a time. Attempting to add another one would result in an exception when attempting to create the subsequent subscription with this type of filter.

**Comment by SeanFeldman** [comment]
What would you expect to be the behaviour for the messages caught previously when their tour is subscribed to later and messages of that type where not confined crib this discussion subscription? For those messages to remain in this special subscription or get purged?

**Comment by mikewo** [comment]
They would remain. I don't expect any specific action to occur if someone adds a subscription that starts getting the messages. The "fall through filter" subscription simply wouldn't get any more of them. It would be up to me to know that we've corrected the issue and then perhaps push messages of that type from the fall through filter back through to correctly process.

Ideally, in this set up the fall through filter subscription should be pretty much empty unless something gets misconfigured, or perhaps an existing filtered subscription gets accidentally deleted. Anything landing in this fall through would then show up as big red flag that something is wrong. You'd react to this, correct the issue, and then know you have to reprocess the items, which could mean just having something resubmit the messages back to the topic, and if the system is corrected then they would get processed as expected (aka, they shouldn't end up back in the fall through subscription).

Likewise, if someone turned on a "trueFilter" for a subscription I would expect this fall through filter to basically get nothing, but any existing messages in it would be left alone. Any other approach would mean that if someone turned on a trueFilter, or simply added a subscription with no defined filter (which is just a true filter really), then all the messages in the fall through would then be gone. I don't want that type of behavior. I feel like it's still important for the fall through subscription to be worked and handled as necessary, just as managing a Dead Letter Queue is an exercise that has to happen.

**Comment by Fabiest** [comment]
I have the same need. Just don't want to lose messages due to some misconfiguration\deletion by mistake.

**Comment by ig-sinicyn** [comment]
I want to up this issue.

We are migrating from RabbitMq to the ServiceBus and we are stuck with lack of troubleshooting features.

There is no detection of routing failures. The legacy api have `TopicDescriptor.EnableFilteringMessagesBeforePublishing` feature but the new one does not.

Also, there is no analog of RabbitMQ's alternate exchange, so there's no way to detect routing failures in production.

**Comment by ig-sinicyn** [comment]
up again:)

**Comment by EldertGrootenboer** [comment]
We have added this to our backlog to investigate the options around this.

**Comment by mbrimble** [comment]
I want this. How do i up vote it?

**Comment by EldertGrootenboer** [comment]
> cc @EldertGrootenboer I want this. How do i up vote it?

You can upvote on the original request with a thumbs up, or leave a response like you did.
If you have any specific input for this request, please also let us know in the comments here.

**Comment by cecilphillip** [comment]
Spend a fair amount of time browsing around for solutions until I landed here. It would be great to have an option to set an "unmatched" subscription or queue that these messages would automatically gets forwarded to, then leave if up to the dev to figure out what to do with them.

**Comment by EldertGrootenboer** [comment]
Thank you for your feedback on this item, it helps us in our efforts to continuously prioritize the different requests we get from all our various feedback channels. We have added this feature in our backlog, however we currently don't have an ETA on when development might start on this. For now, to help us give this the right priority, it would be helpful to see others vote and support this feature, as well as explain their scenarios.

**Comment by EldertGrootenboer** [comment]
This item in our backlog, however we currently don't have an ETA on when development might start on this. For now, to help us give this the right priority, it would be helpful to see others vote and support this item.

**Comment by Kryptos-FR** [comment]
This is a critical feature. How can we make it happen?

We had some data loss because of a misconfiguration. This feature would have helped us fix the configuration and replay the messages easily. Instead, we had a painful manual process.

**Comment by Kryptos-FR** [comment]
@EldertGrootenboer 20+ up votes (as of 2024-11-14 10:50+00) is not enough? What's the required threshold for your team to act on it?

It's hard to get more votes because the issue usually arises after some production crisis and people are usually busy fixing that before looking for an existing GitHub issue.

**Comment by mbrimble** [comment]
I am sorry I have retired and no longer work. Good luck with getting this issue remedied.

Sent from my iPhone

On 14 Nov 2024, at 11:52 PM, Nicolas Musset ***@***.***> wrote:

﻿

@EldertGrootenboer<https://github.com/EldertGrootenboer> 20+ up votes (as of 2024-11-14 10:50+00) is not enough? What's the required threshold for your team to act on it?

It's hard to get more votes because the issue usually arises after some production crisis and people are usually busy fixing that before looking for an existing GitHub issue.

**Comment by SeanFeldman** [comment]
> This item is in the upper regions of our backlog

Could you share a visualization of your backlog structure and breakdown for visual folks? An "upper region" could be in the "lower section" of one of the "stages" 😃 

I'm just kidding, but there's a bit that's serious.

**Comment by zodapopinski** [comment]
We need the **EnableFilteringMessagesBeforePublishing** property as well. Just fell into this one as well while migrating to newer libraries.

**Comment by janders088** [comment]
Not sure if this helps anyone, but it appears this property can still be set by using ServiceBusExplorer

<img width="749" height="415" alt="Image" src="https://github.com/user-attachments/assets/efb20c0d-1544-4925-aa43-0edc062d2fce" />

I just tested it by sending a message from a Logic App to a Topic configured with this property set to true and received the expected error message.

<img width="1591" height="633" alt="Image" src="https://github.com/user-attachments/assets/51cfb419-7efe-4d5f-a36e-7a0fb7c4e5e1" />

**Comment by Kryptos-FR** [comment]
@janders088 did the message go to the dead-letter queue?

**Comment by SeanFeldman** [comment]
> Not sure if this helps anyone, but it appears this property can still be set by using ServiceBusExplorer

@janders088 SBE is leveraging the legacy SDK.

**Comment by janders088** [comment]
@SeanFeldman Hi Sean, it does indeed but the fact remains that Azure Service Bus still recognises the **EnableFilteringMessagesBeforePublishing** being configured to true\false. What I am trying to ascertain next is if this property can also be set using Bicep \ ARM Templates, which I honestly doubt will be possible but would like to know for sure.

**Comment by janders088** [comment]
@Kryptos-FR no, it simply returned a 400 status code and 'no matching subscription found' error message to the process attempting to publish the message, which in this instance was a Logic App workflow.

**Comment by Kryptos-FR** [comment]
So this doesn't fix the issue. Message is still lost.

**Comment by janders088** [comment]
@Kryptos-FR no, but in theory, I guess you could employ application-level dead-lettering to explicitly dead letter a message after receiving the 'no matching subscription found error'

**Comment by Kryptos-FR** [comment]
That only works if you are the one sending the message, when it's a third-party, they might not be able to change their logic (or for a free), and if you introduced an intermediate azure function for instance that is just passthrough with this retry logic, you are paying for that function with all that it implies while adding yet another failure point that you need to monitor. All for a feature that should be native at the service bus level.

**Comment by janders088** [comment]
@Kryptos-FR I completely agree that this should be a native Service Bus feature. I was simply trying to offer a potential workaround in the meantime, as the backlog containing this feature request doesn't seem to be moving very quickly (if at all)

**Comment by dobbean** [comment]
Please add this feature request. I'm amazed that after all this time this is still not implemented. This should have existed from the get-go.
